### PR TITLE
Harvest performance fee

### PR DESCRIPTION
**Pull request by ryuheimat**
To keep TVL of vault, we decided to add harvesting performance fee feature instead of transfer at redeem operation.

### CODE REVIEW

**Comment by 0xMikazuki** [comment]
y build failures?

```
Error: test/strategy/anchor/AnchorNonUSTStrategy.spec.ts(87,32): error TS2554: Expected 5-6 arguments, but got 4.
Error: test/strategy/anchor/AnchorNonUSTStrategy.spec.ts(110,7): error TS2554: Expected 9-10 arguments, but got 11.
Error: test/strategy/anchor/AnchorNonUSTStrategy.spec.ts(156,34): error TS2554: Expected 5-6 arguments, but got 4.
Error: test/strategy/anchor/AnchorNonUSTStrategy.spec.ts(181,29): error TS2339: Property 'treasury' does not exist on type 'AnchorNonUSTStrategy'.
Error: test/strategy/anchor/AnchorNonUSTStrategy.spec.ts(192,29): error TS2339: Property 'perfFeePct' does not exist on type 'AnchorNonUSTStrategy'.
Error: test/strategy/anchor/AnchorNonUSTStrategy.spec.ts(289,29): error TS2339: Property 'convertedUst' does not exist on type 'AnchorNonUSTStrategy'.
```

**Comment by ryuheimat** [comment]
Yeah, This is draft for now, not completed.
Unit tests will be updated soon.

**Comment by gabrielpoca** [comment]
@ryuheimat I'm a bit confused by this. I thought we would only add performance fee to gains AKA yield, no?

File: contracts/Vault.sol [hunk 1]
```diff
@@ -46,6 +46,7 @@ contract Vault is
     //
 
     bytes32 public constant INVESTOR_ROLE = keccak256("INVESTOR_ROLE");
+    bytes32 public constant HARVESTOR_ROLE = keccak256("HARVESTOR_ROLE");
```
Comment (gabrielpoca) [inline contracts/Vault.sol:49]: What would happen if we let anyone call the harvest function?

Comment (ryuheimat) [reply contracts/Vault.sol:49]: the reason we decided to harvest fee is that we keep fee in contract, not immediately send to treasury.
if there is no permission, anybody can claim fee to reduce tvl.

**Comment by ryuheimat** [comment]
> @ryuheimat I'm a bit confused by this. I thought we would only add performance fee to gains AKA yield, no?

no, even we didn't redeem, we should calculate performance fee to calculate new deposit's share correctly.

**Comment by gabrielpoca** [comment]
@ryuheimat maybe I'm misunderstanding, but my understanding was that the performance fee is applied to the profit AKA yield. Why would you apply a fee for sponsoring?

File: contracts/Vault.sol (hunk 1 above)
Comment (gabrielpoca) [reply contracts/Vault.sol:49]: I see. It makes sense; you can keep the money around 👍  about the role, why not use the INVESTOR_ROLE? that's already used to control when the money is invested

**Comment by ryuheimat** [comment]
> @ryuheimat maybe I'm misunderstanding, but my understanding was that the performance fee is applied to the profit AKA yield. Why would you apply a fee for sponsoring?

this does not take fee from sponsor.
It only takes fee from yield, and sponsor does not generate yield.

File: contracts/Vault.sol (hunk 1 above)
Comment (ryuheimat) [reply contracts/Vault.sol:49]: we could use INVESTOR_ROLE, however to make sure roles, i added another role.

Comment (naps62) [reply contracts/Vault.sol:49]: I don't understand. the fee is always sent to `treasury`. it doesn't matter who calls the function

File: contracts/Vault.sol [hunk 2]
```diff
@@ -228,7 +273,7 @@ contract Vault is
     function claimYield(address _to) external override(IVault) nonReentrant {
         require(_to != address(0), "Vault: destination address is 0x");
 
-        uint256 yield = yieldFor(_msgSender());
+        uint256 yield = _yieldForWithPerfFee(_msgSender());
```
Comment (gabrielpoca) [inline contracts/Vault.sol:276]: What if we just changed the `yieldFor` function to return `(uint256 yield, uint256 perfFee)` instead? I

File: contracts/Vault.sol [hunk 3]
```diff
@@ -358,24 +407,35 @@ contract Vault is
         _unsponsor(_to, _ids);
     }
 
+    function harvest() external onlyRole(HARVESTOR_ROLE) {
```
Comment (gabrielpoca) [inline contracts/Vault.sol:410]: I have to be honest, I don't like the name "harvest", to be "harvest" and "claim" feel like the same thing. Can't we rename this to "withdrawPerformanceFee"?

File: contracts/Vault.sol [hunk 4]
```diff
@@ -663,9 +723,9 @@ contract Vault is
     function _transferAndCheckUnderlying(address _from, uint256 _amount)
         internal
     {
-        uint256 balanceBefore = totalUnderlying();
+        uint256 balanceBefore = underlying.balanceOf(address(this));
```
Comment (gabrielpoca) [inline contracts/Vault.sol:726]: I'm lost, what's this change about?

File: contracts/mock/MockERC20.sol [hunk 5]
```diff
@@ -22,4 +22,8 @@ contract MockERC20 is ERC20("MockERC20", "ME2") {
     function mint(address _user, uint256 _amount) public {
         _mint(_user, _amount);
     }
+
+    function burn(address _user, uint256 _amount) public {
```
Comment (gabrielpoca) [inline contracts/mock/MockERC20.sol:26]: Where are you using this burn?

**Review by gabrielpoca — COMMENTED** [review]
One question, for the `accumulatedPerfFee`, do we want to have that money invested until we take it out, or should we set it aside?

File: contracts/Vault.sol (hunk 4 above)
Comment (naps62) [reply contracts/Vault.sol:726]: I guess it's an equivalent way of checking the amount, but without doing unnecessary cross-contract calls to the strategy

Comment (ryuheimat) [reply contracts/Vault.sol:726]: yes

**Comment by ryuheimat** [comment]
> One question, for the `accumulatedPerfFee`, do we want to have that money invested until we take it out, or should we set it aside?

We can keep accumulatedPerfFee as invested - something similar with sponsor.

File: contracts/Vault.sol (hunk 2 above)
Comment (ryuheimat) [reply contracts/Vault.sol]: updated

File: contracts/Vault.sol (hunk 3 above)
Comment (ryuheimat) [reply contracts/Vault.sol]: fixed

**Comment by gabrielpoca** [comment]
> > One question, for the `accumulatedPerfFee`, do we want to have that money invested until we take it out, or should we set it aside?
> 
> We can keep accumulatedPerfFee as invested - something similar with sponsor.

I guess so, but what's the ideal scenario? Did you check with @CrisBRM ?

**Comment by gabrielpoca** [comment]
@CrisBRM we'll have to revisit that use-case later on :/ it's a bit hard to take that into account at this point. But what I get from your message is that we want to keep the performance fee around to boost yield, which means @ryuheimat implementation is correct 👍

**Comment by CrisBRM** [comment]
> @CrisBRM we'll have to revisit that use-case later on :/ it's a bit hard to take that into account at this point. But what I get from your message is that we want to keep the performance fee around to boost yield, which means @ryuheimat implementation is correct 👍

For clarity, the performance fee should belong to the treasury but the yield generated by it should go to other depositors. Whether we should withdraw and redeposit or it can be automagically assigned as sponsor is up to you.

**Comment by gabrielpoca** [comment]
In this scenario the performanceFee is in the vault but can only be transferred to the treasury. It's only transferred when we chose to do it. While it's in the vault, it's generating yield which is automatically distributed by everyone. 👍

File: test/Vault.spec.ts [hunk 6]
```diff
@@ -826,36 +1016,32 @@ describe("Vault", () => {
...
           carol.address,
-          parseUnits("50"),
-          parseUnits("25").mul(SHARES_MULTIPLIER)
+          yieldWithFee.sub(perfFee),
+          parseUnits("25").mul(SHARES_MULTIPLIER),
+          perfFee
         );
     });
 
-    it("claims the yield of a user", async () => {
```
Comment (gabrielpoca) [inline test/Vault.spec.ts:839]: Why did you remove the test to emit an event? We had two tests instead of one to make it easier to debug if something was wrong.

File: test/Vault.spec.ts [hunk 7]
```diff
@@ -905,10 +1094,19 @@ describe("Vault", () => {
       });
 
       await vault.connect(alice).deposit(params);
-      await addYieldToVault("100");
-
-      expect(await vault.yieldFor(alice.address)).to.eq(parseUnits("50"));
-      expect(await vault.yieldFor(bob.address)).to.eq(parseUnits("50"));
+      const newYield = await addYieldToVault("100");
```
Comment (gabrielpoca) [inline test/Vault.spec.ts:1097]: I guess this variable could be just `yield`?

File: test/Vault.spec.ts (hunk 6 above)
Comment (ryuheimat) [reply test/Vault.spec.ts:839]: i think we don't need add one test case for testing event.
I added event test in next case.

File: test/Vault.spec.ts (hunk 7 above)
Comment (ryuheimat) [reply test/Vault.spec.ts:1097]: It is a new generated yield.
yield is a total yield.
In this test case, we added yield one time, so newYield = total yield, however newYield is more make sense to me.

File: contracts/Vault.sol [hunk 8]
```diff
@@ -186,7 +225,11 @@ contract Vault is
         public
         view
         override(IVault)
-        returns (uint256)
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
```
Comment (naps62) [inline contracts/Vault.sol:232]: can we please name these return fields? (either here, on in the function docs)
It's hard to understand what each one means

File: contracts/vault/IVault.sol [hunk 9]
```diff
@@ -110,12 +115,19 @@ interface IVault {
      *
      * @param _to address to consider.
      *
-     * @return amount of yield for @param _to.
+     * @return amount of yield for @param _to, share of yield, and performance fee from yield
```
[the remark this answers is not in the document]
Comment (ryuheimat) [reply contracts/vault/IVault.sol:118]: hmm, 
`@return the yield, shares, and performance fee for @param _to` this does not make sense to me.
because performance fee is from yield, not `@param _to(user)`.

File: contracts/Vault.sol (hunk 8 above)
Comment (ryuheimat) [reply contracts/Vault.sol:232]: `@return amount of yield for @param _to, share of yield, and performance fee from yield`, updated description in IVault.sol

**Comment by naps62** [comment]
@ryuheimat can you check the encoding of the USTStrategy and NonUSTStrategy?

The newline encoding seems to be in windows format, causing the diff to show the entire file

image

File: test/Vault.spec.ts [hunk 10]
```diff
@@ -152,24 +158,219 @@ describe("Vault", () => {
...
+          TWO_WEEKS,
+          INVEST_PCT,
+          TREASURY,
+          constants.AddressZero,
+          PERFORMANCE_FEE_PCT
+        )
+      ).to.be.revertedWith("Vault: owner cannot be 0x0");
+    });
+
+    it("Check initial values", async () => {
```
[the remark this answers is not in the document]
Comment (ryuheimat) [reply test/Vault.spec.ts:248]: yeah. check initial values set in constructor.

**Comment by ryuheimat** [comment]
Yeah, I switched new PC(with windows OS), and shows as entire file changed,
Not sure how i can solve it.

File: test/Vault.spec.ts (hunk 7 above)
Comment (ryuheimat) [reply test/Vault.spec.ts:1097]: and `yield` is a typescript keyword which we cannot use as variable

**Comment by naps62** [comment]
On windows you can configure git to still use linux-style newlines when commiting files (and should still convert to windows format locally)

https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

Please do that as it's very hard to review the files right now, since it's impossible to see what exactly changed

File: contracts/vault/IVault.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply contracts/vault/IVault.sol]: ok, then something different maybe? If you don't have any ideas I can suggest something more verbose:

```
@return yield for @param _to, the performance fee for that yield, and the shares for the yield plus the performance fee.
```

File: test/Vault.spec.ts (hunk 6 above)
Comment (gabrielpoca) [reply test/Vault.spec.ts]: We could write all the tests in a single test case, but doing so makes it harder to read and understand what's failing. That's why I wrote two test cases, with one just for the event. Maybe only one of them fails, and that immediately sends a message about what could be wrong. The test was already there, and I don't see why you would remove it.

**Comment by ryuheimat** [comment]
@naps62 issue has been fixed, you can now check changes.

**Review by naps62 — APPROVED** [review]
